neps/nep-0000.md

```diff
@@ -0,0 +1,215 @@
+
+
+## Specification
+
+### Guide-level explanation
+
+We should be able to do the following:
+- Approve a spender to withdraw funds on behalf of the balance owner account
+- Allow spender to either be an Account or an Access Key
+- Allow spender to withdraw funds
+- Allow an account to get their approvals
+
+There are a few concepts in the scenario above that needs explaining.
+
+- **Spender**: An account or an access key that has approval to withdraw funds from another account
+-  **Balance owner account**: An account that is the subject of the withdrawal. 
+-  **Allowance**: The amount of tokens for the spender account that the spender can use on behalf of the balance account owner.
+-  **Approval**: A mapping of Account Owner and Spender to allowance.
+
+
+## Reference Implementation
+
+**NOTES**
+
+- All amounts, balances and allowance are limited by U128 (max value 2**128 - 1).
+- Having `0` as amount is not permitted, unlike ERC20 specification.
+- There is a need to explicitly mention `ft_transfer_from_account` and `ft_transfer_from_key` due to an edge case where giving a function call access key to call `ft_transfer_from` would allow this access key to use any of the funds approved for the contract account id itself. Thus, `ft_transfer_from` is divided into two different functions.
+
+```rs
+    // ---- Approval standard -----
+
+    /// Enum that is used to identify the Spender
+    pub enum AccountOrKey {
+        Account(AccountId),
+        Key(PublicKey),
+    }
+
+    /// Approve the passed address to spend the specified amount of tokens on behalf of
+    /// `env::predecessor_account_id`.
+    ///
+    /// Arguments:
+    /// * `spender` The address which will spend the funds.
+    /// * `current_value` The amount of tokens currently allowed. This is used to ensure atomicity.
+    /// * `value` The amount of tokens to be spent.
+    pub fn ft_approve(&mut self, spender: AccountOrKey, current_value: U128, value: U128)
+
+    /// Transfer funds from a `from` to `to` specifically for `account` as spender
+    /// * `to` The address which the transfer will be made from
+    /// * `from` The address which the transfer will be made to
+    /// * `amount' The amount of tokens to be transferred
+    /// * `memo` Optional message to be sent with the transfer
+    pub fn ft_transfer_from_account(
+        &mut self,
+        from: AccountId,
+        to: AccountId,
+        amount: U128,
+        memo: Option<String>,
+    )
+     /// Transfer funds from a `from` to `to` specifically for `key` as spender
+    /// * `from` The address which the transfer will be made from
+    /// * `to` The address which the transfer will be made to
+    /// * `amount' The amount of tokens to be transferred
+    /// * `memo` Optional message to be sent with the transfer
+    pub fn ft_transfer_from_key(
+        &mut self,
+        from: AccountId,
+        to: AccountId,
+        amount: U128,
+        memo: Option<String>,
+    )
```

Comment:
> there is no way to know at runtime, which key was used for signing the transaction

@mfornet Well, there is one - [`near_sdk::env::signer_account_pk()`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.signer_account_pk.html), it was heavily used by the old `multisig` contract. Yet, the signer key/account id MUST NOT be used for the validation as this would open up vulnerabilities where a user makes a call to a malicious contract which makes a cross-contract call to this method which checks the signer account and suddenly authorizes some actions on behalf of the signer account while the user only made a call to the malicious account.

Reply:
Correct, multisig and this kind of methods where this is done for initiating an action with access key should be checking `predecessor_account_id == signer_account_id` if the key check is indeed being applied.